### ISSUE TITLE
incorrect mask with additionalProperties:true and non-required properties

### ISSUE THREAD

**Issue by mmoskal** [mlc-ai/xgrammar#208]
With the following schema, xgrammar will accept `{"a": "wrong"}` even though it shouldn't ("a" is supposed to be an integer)

```json
{
    "type": "object",
    "properties": {
        "a": {"type": "integer"},
    },
    "additionalProperties": True,
    "required": [],
}
```

full repro

```python
import xgrammar as xgr
import torch
import numpy as np
import json

from transformers import AutoTokenizer, AutoConfig
# Get tokenizer info
model_id = "unsloth/Meta-Llama-3.1-8B-Instruct"
tokenizer = AutoTokenizer.from_pretrained(model_id)
config = AutoConfig.from_pretrained(model_id)
# This can be larger than tokenizer.vocab_size due to paddings
full_vocab_size = config.vocab_size
tokenizer_info = xgr.TokenizerInfo.from_huggingface(tokenizer, vocab_size=full_vocab_size)

compiler = xgr.GrammarCompiler(tokenizer_info, max_threads=1)

schema = {
    "type": "object",
    "properties": {
        "a": {"type": "integer"},
    },
    "additionalProperties": True,
    "required": [],
}
compiled_grammar = compiler.compile_json_schema(json.dumps(schema))
matcher = xgr.GrammarMatcher(compiled_grammar)
token_bitmask = xgr.allocate_token_bitmask(1, tokenizer_info.vocab_size)

sim_sampled_response = '{"a": "wrong"}'
sim_sampled_token_ids = tokenizer.encode(sim_sampled_response, add_special_tokens=False)

for i, sim_token_id in enumerate(sim_sampled_token_ids):
    matcher.fill_next_token_bitmask(token_bitmask)
    assert matcher.accept_token(sim_token_id)
```

**Comment by mmoskal** [comment]
This of course affects great many testcases if you enable `strict=False` (which is to say, JSON-Schema compliant mode).

**Comment by ianliuy** [comment]
## Analysis & Fix Plan for #208

### Root Cause

In `GetPartialRuleForProperties()` (`cpp/json_schema_converter.cc`), the additional property pattern uses `GetKeyPattern()` which returns `basic_string` — matching **any** JSON string key, including defined property names. When this catch-all alternative is added to the grammar (lines ~1642, ~1745, ~1870 for the three grammar generation cases), the parser tracks both the typed path and the catch-all path simultaneously. The token bitmask becomes the **union** of valid tokens across all paths, so the catch-all's `any_rule` overrides the typed constraint.

### Fix Approach

Replace `GetKeyPattern()` with a new **trie-based key exclusion pattern** (`GetKeyPatternExcluding()`) that matches any JSON string key **except** the defined property names. This ensures defined keys can only match via their typed rule, while truly unknown keys still use the catch-all with `any_rule`.

Affected call sites: lines ~1603, ~1659, ~1770 (all 3 grammar generation cases in `GetPartialRuleForProperties`).

### Regression Test

Adding test cases in `tests/python/test_json_schema_converter.py` for:
- `additionalProperties: true` + empty `required` + typed property → reject wrong type
- Same schema → accept correct type
- Same schema → accept truly additional (unknown) key with any value

I'm working on a fix.

**Comment by ianliuy** [comment]
Implemented as described above. PR: https://github.com/mlc-ai/xgrammar/pull/595

The fix replaces `GetKeyPattern()` (which matched ANY key) with `GetKeyPatternExcluding()` (which excludes defined property names) at 3 call sites in `GetPartialRuleForProperties()`. Added 7 regression test cases.
